src/dec128.rs

```diff
@@ -136,6 +136,79 @@ impl From<u32> for d128 {
     }
 }
 
+/// based on wiki's algorithm.
+fn double_dabble(num: u64, res: &mut [u8]) {
+    let nscratch = res.len();
+
+    let mut smin = nscratch - 2;    /* speed optimization */
+
+    for j in 0..64 {
+        /* This bit will be shifted in on the right. */
+        let shifted_in = if (num & (1u64 << (63 - j))) > 0 { 1 } else { 0 };
+
+        /* Add 3 everywhere that scratch[k] >= 5. */
+
+        for k in smin..nscratch {
+            res[k] += if res[k] >= 5 { 3 } else { 0 };
+        }
+
+        /* Shift scratch to the left by one position. */
+        if res[smin] >= 8 {
+            smin -= 1
+        }
+
+        for k in smin..nscratch - 1 {
+            res[k] <<= 1;
+            res[k] &= 0xF;
+            res[k] |= if res[k + 1] >= 8 { 1 } else { 0 };
+        }
+
+
+        /* Shift in the new bit from arr. */
+        res[nscratch - 1] <<= 1;
+        res[nscratch - 1] &= 0xF;
+        res[nscratch - 1] |= shifted_in;
+    }
+}
+
+/// Converts an i64 to d128. The result is exact and no error is possible.
+impl From<i64> for d128 {
```

Comment:
Conversion is expensive enough so this can be simply be coded as:

```rust
if val < 0 { d128::from((~(val as u64) + 1)) } else { d128::from(val as u64) }
```